### PR TITLE
Add cli flag '--disable-hot-restart'.

### LINKED ISSUE #2029: Add CLI flag to disable hot restart
Hot Restart should be disable-able via a CLI argument

> mklein [4:28 PM] 
not currently, we could add a CLI option to disable hot restart
[4:29] 
please file an issue on that

https://envoyproxy.slack.com/archives/C78M4KW76/p1510187339000151

### PR DESCRIPTION
**Pull request by tonya11en**
hot restart: add cli flag to disable hot restart functionality

*Description*:
>This PR attempts to fix #2029 by adding a command line flag to disable hot restart. This can still be accomplished via the currently existing compile flag.

*Risk Level*: Low | Medium | High
>low

*Testing*:
> Unit testing

*Docs Changes*:
> N/A

*Release Notes*:
>TODO

Fixes #2029 

Signed-off-by: Tony Allen <tallen@lyft.com>

### CODE REVIEW

**Comment by mattklein123** [comment]
Thanks for working on this! Some ultra quick notes: 
1) This is going to conflict with https://github.com/envoyproxy/envoy/pull/2568. Please let that land first and potentially work with @jmarantz and @dnoe on this.
2) Instead of checking that hot restart is disabled in the real hot restart implementation, the flag should instead using the NOP hot restarter like is currently done via ifdef in main.

**Comment by jmarantz** [comment]
Thanks @mattklein123.  Actually having --disable_hot_restart as a flag would be great.  I hadn't added a flag before and I wasn't sure about all the steps required.  Tony -- what do you think of splitting this PR so you just add the flag to Options/OptionsImpl here, but not make it operational.  That wouldn't conflict with my PR and I could merge and use your flag in lieu of some explicit args I added.

Also, does this need a doc update associated with it?

**Comment by tonya11en** [comment]
@jmarantz sounds good, I'll just add the flag. Thanks for taking a look!

**Comment by tonya11en** [comment]
@mattklein123 is there anything I should do on my end to update the documentation? I grepped around the repo and didn't see files that contained the linked cli docs.

**Comment by mattklein123** [comment]
Documentation is in the data-plane-api repo. See here: https://github.com/envoyproxy/data-plane-api/blob/master/CONTRIBUTING.md

File: test/mocks/server/mocks.cc [hunk 1]
```diff
@@ -28,6 +28,9 @@ MockOptions::MockOptions(const std::string& config_path)
   ON_CALL(*this, logPath()).WillByDefault(ReturnRef(log_path_));
   ON_CALL(*this, maxStats()).WillByDefault(Return(1000));
   ON_CALL(*this, maxObjNameLength()).WillByDefault(Return(150));
+  ON_CALL(*this, hotRestartDisabled()).WillByDefault(Invoke([this] {
```
Comment (mattklein123) [inline test/mocks/server/mocks.cc]: I would use `ReturnPointee` here.

**Comment by tonya11en** [comment]
I'll figure out what's going on with that ReturnPointee call one night this week. Can't do much during the work day.

**Comment by htuch** [comment]
@tonya11en Now that https://github.com/envoyproxy/envoy/pull/2568 is merged, do you want to merge master and make any needed changes?

**Comment by tonya11en** [comment]
@jmarantz once this gets approved/merged, should I just go ahead and work it into your recently merged PR? Not sure if you had other plans regarding this.

**Comment by tonya11en** [comment]
@htuch sounds good to me, I just need someone to approve this PR. I'm not authorized to merge this guy.

**Pull request merged**